File: code/Scala/Sorting.scala
import scala.compiletime.ops.string
import scala.io.Source

object Sorting {
    def selectionSort(arr: Array[Int], n: Int) = {
        // println("Array before sorting: ");
        // printArray(arr);
        
        val startTime = System.currentTimeMillis();

        for (i <- 0 until n) {
            var minIndex = i;
            for(j <- i until n if arr(j) < arr(minIndex)) minIndex = j
            val temp = arr(i);
            arr(i) = arr(minIndex);
            arr(minIndex) = temp;
        }

        val endTime = System.currentTimeMillis();
        val elapsedTime = (endTime - startTime) / 1000F;

        // println("Array after sorting: ");
        // printArray(arr);

        println("Selection sort in Scala took: " + elapsedTime + " seconds");
    }

    def printArray(arr: Array[Int]) = {
        for(a <- arr) print(s"$a ")
    }

    def readFile(filename: String, arr: Array[Int]) = {
        val bufferedSource = Source.fromFile(filename);
        val numbers: Array[String] = bufferedSource.getLines.mkString.split(", ");
        bufferedSource.close;
        for(i <- 0 until numbers.length) arr(i) = numbers(i).toInt
    }

    def main(args: Array[String]) = {
        var arr = new Array[Int](10000);
        readFile("/Users/abhigupta/Development/benchmarking/data/numbers.txt", arr);
        selectionSort(arr, arr.length);
    }
}

File: code/JavaScript/sorting.js
const { start } = require('repl');

function selectionSort(arr, n)
{
    // console.log("Array before sorting: #{arr}")
    // printArray(arr, n);
    var startTime, endTime;

    startTime = new Date();

    var i, j, minIndex;
    for(i = 0; i < n; i++) {
        minIndex = i;
        for(j = i; j < n; ++j) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        temp = arr[i];
        arr[i] = arr[minIndex];
        arr[minIndex] = temp;
    }

    endTime = new Date();
    var execution_time = ((endTime - startTime)/1000).toFixed(3);

    // console.log("Array after sorting: #{arr}")
    // printArray(arr, n);
    console.log("Selection sort in JavaScript took: " + execution_time + " seconds")
}

function readFile(filename)
{
    var fs = require('fs');
    data = fs.readFileSync(filename).toString().split(',');
    return data.map(function (x) { 
        return parseInt(x, 10); 
    });
}

function printArray(arr) 
{
    for(i = 0; i < arr.length; ++i) {
        process.stdout.write(arr[i] + " ");
    }
    console.log();
}

arr = readFile("/Users/abhigupta/Development/benchmarking/data/numbers.txt")
selectionSort(arr, arr.length)